feat!: aztec-nargo now directly exposes nargo. aztec-postprocess-contract transpiles and generates vks

dbanks12 opened the pull request:
## What's done and not done?
1. ✅ Make aztec-nargo just be a direct exposure of the version of nargo in the container. To this end aztec-nargo should from the users experience, just be plain vanilla nargo, but just versioned as we need.
1. ✅ Fix the version number in aztecs nargo build to just be whatever our current noir-ref is, maybe with a -aztec suffix or something to ensure we have more context.
    - Setting the `GIT_COMMIT` env var to the current noir-ref with `-aztec` appended.
        ```
        % aztec-nargo --version
        nargo version = 1.0.0-beta.9
        noirc version = 1.0.0-beta.9+nightly-2025-08-04-aztec
        (git version hash: nightly-2025-08-04-aztec, is dirty: false)
        ```
        - **Note that we are appending `-aztec` to `GIT_COMMIT`, which does not end up in the "nargo version", but appears in the noirc version and git version hash.** I don't see a clean/simple way to append `-aztec` to the reported nargo version.
1. ✅  Put the current compile_and_postprocess script behind a command e.g. `aztec-postprocess-contract` , make it no longer run nargo, but rather just transform the artifact with the transpiler and vk generation.
1. **[TODO]** ⚠️ Ensure devrel update docs promptly to reflect the added compile step for aztec.

## An important note on caching and recompiling
1. The transpiler won't re-transpile a contract that is already transpiled. *Yay!*
3. VK caching is still used. *Yay!*
4. Buuuuut... we can no longer use automatically usee nargo's `--show-artifact-paths` to inform what contracts have actually changed and therefore must be post-processed. So, the "update the artifact sequentially with each generated vk" phase runs for every artifact no matter what. If the source hasn't changed, the result will be the same, but the files will be touched.

dbanks12 commented:
* **#16172** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16172?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16172?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sirasistant commented:
The version and commit hash comes from these env variables:
```rust
const GIT_HASH: &str = env!("GIT_COMMIT");
const IS_DIRTY: &str = env!("GIT_DIRTY");
const NARGO_VERSION: &str = env!("CARGO_PKG_VERSION");

static VERSION_STRING: &str = formatcp!(
    "version = {}\nnoirc version = {}\n(git version hash: {}, is dirty: {})",
    NARGO_VERSION,
    NOIR_ARTIFACT_VERSION_STRING,
    GIT_HASH,
    IS_DIRTY
);
```

I think we need to update the noir/bootstrap script to use the commit of the synced noir version
```bash
# Fake this so artifacts have a consistent hash in the cache and not git hash dependent.
export GIT_COMMIT="0000000000000000000000000000000000000000"
export SOURCE_DATE_EPOCH=0
export GIT_DIRTY=false
export RUSTFLAGS="-Dwarnings"
```

sirasistant commented:
> Are we sure we want to call it aztec-compile-contract? I think of it more as post-processing. Nargo is responsible for compiling.

maybe aztec-contract-postprocess ?

sirasistant reviewed (approved):
Changes LGTM, but we need to change commit hash in bootstrap too

critesjosh commented on docs/docs/developers/guides/js_apps/deploy_contract.md:
````diff
@@ -23,6 +23,7 @@ Compile the contract:
 
 ```bash
 aztec-nargo compile
+aztec-compile-contract
````
both of these steps are needed now? the naming is confusing. based on the name here i would expect these steps to be doing the same thing

dbanks12 replied:
Yes, both are needed. `aztec-postprocess-contract` was also suggested. The second step transpiles public functions and creates VKs for private functions. I'm happy to change to postprocess unless you have any other ideas?

critesjosh commented on docs/docs/developers/guides/smart_contracts/how_to_compile_contract.md:
```diff
@@ -12,18 +12,29 @@ We'll also cover how to generate a helper [TypeScript interface](#typescript-int
-This will output a JSON artifact for each contract in the project to a `target` folder containing the Noir ABI artifacts.
+This will output JSON artifacts for each contract in the project to a `target` folder.
+
+2. Then, run the postprocessing step to prepare contracts for Aztec:
```
Based on this addition, what about calling it `aztec-postprocess-contract`?

dbanks12 replied:
Ah, yes! That was the other suggestion. Let's do that.

critesjosh commented on docs/docs/developers/guides/js_apps/deploy_contract.md:
````diff
@@ -24,6 +24,7 @@ Compile the contract:
 
 ```bash
 aztec-nargo compile
````
```suggestion
aztec-nargo compile # generate contract artifacts
```

critesjosh commented on docs/docs/developers/guides/js_apps/deploy_contract.md:
````diff
@@ -24,6 +24,7 @@ Compile the contract:
 
 ```bash
 aztec-nargo compile
+aztec-postprocess-contract
````
```suggestion
aztec-postprocess-contract # transpile contract and generate verification keys
```

critesjosh commented on docs/docs/developers/tutorials/codealong/contract_tutorials/counter_contract.md:
````diff
@@ -125,19 +125,20 @@ This is a `utility` function which is used to obtain the counter information out
+In `./counter/` directory, run these commands:
 
 ```bash
 aztec-nargo compile
````
```suggestion
aztec-nargo compile # generate contract artifacts
```

critesjosh commented on docs/docs/developers/tutorials/codealong/contract_tutorials/counter_contract.md:
````diff
@@ -125,19 +125,20 @@ This is a `utility` function which is used to obtain the counter information out
 
 ```bash
 aztec-nargo compile
+aztec-postprocess-contract
````
```suggestion
aztec-postprocess-contract # transpile contracts and generate verification keys
```

critesjosh reviewed (approved):
docs changes lgtm

TomAFrench reviewed (approved):
patch looks good

dbanks12 committed:
feat!: aztec-nargo now directly exposes nargo. aztec-postprocess-contract transpiles and generates vks

## What's done and not done?
1. ✅ Make aztec-nargo just be a direct exposure of the version of nargo in the container. To this end aztec-nargo should from the users experience, just be plain vanilla nargo, but just versioned as we need.
1. ✅ Fix the version number in aztecs nargo build to just be whatever our current noir-ref is, maybe with a -aztec suffix or something to ensure we have more context.
    - Setting the `GIT_COMMIT` env var to the current noir-ref with `-aztec` appended.
        ```
        % aztec-nargo --version
        nargo version = 1.0.0-beta.9
        noirc version = 1.0.0-beta.9+nightly-2025-08-04-aztec
        (git version hash: nightly-2025-08-04-aztec, is dirty: false)
        ```
        - **Note that we are appending `-aztec` to `GIT_COMMIT`, which does not end up in the "nargo version", but appears in the noirc version and git version hash.** I don't see a clean/simple way to append `-aztec` to the reported nargo version.
1. ✅  Put the current compile_and_postprocess script behind a command e.g. `aztec-postprocess-contract` , make it no longer run nargo, but rather just transform the artifact with the transpiler and vk generation.
1. **[TODO]** ⚠️ Ensure devrel update docs promptly to reflect the added compile step for aztec.

## An important note on caching and recompiling
1. The transpiler won't re-transpile a contract that is already transpiled. *Yay!*
3. VK caching is still used. *Yay!*
4. Buuuuut... we can no longer use automatically usee nargo's `--show-artifact-paths` to inform what contracts have actually changed and therefore must be post-processed. So, the "update the artifact sequentially with each generated vk" phase runs for every artifact no matter what. If the source hasn't changed, the result will be the same, but the files will be touched.
